lib/exercises.rb

```diff
@@ -1,22 +1,54 @@
-
-
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
-
-def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+# Time Complexity: O(n) + sort
+# Space Complexity: O(n)
+def grouped_anagrams(strings) 
+  hash = {}
+  
+  strings.each do |string|   
+    # to avoid collision
+    key = string.chars.sort.join
+    
+    if hash[key].nil?
+      hash[key] = [string]
+    else
+      hash[key] << string
+    end 
+  end 
+  
+  return hash.values
 end
 
+##pp grouped_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
+
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n+m)
+# Space Complexity: O(n)
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
-end
+  hash = {}
+  result = []
+  
+  list.each do |num|
+    if hash[num].nil?
+      hash[num] = 1
+    else
+      hash[num] += 1
+    end
+  end
+  
+  i = 1
+  hash.each do |key, value|
+    if i <= k
+      result << key
+      i += 1
+    end 
+  end
```

Comment:
I'm not sure how this works because how do you know the 1st K elements of the hash are the most frequent ones.